packages/react-dom-bindings/src/server/ReactDOMServerExternalRuntime.js

```diff
@@ -13,13 +13,65 @@ import {
   completeSegment,
 } from './fizz-instruction-set/ReactDOMFizzInstructionSet';
 
-// Intentionally does nothing. Implementation will be added in future PR.
-// eslint-disable-next-line no-unused-vars
-const observer = new MutationObserver(mutations => {
-  // These are only called so I can check what the module output looks like. The
-  // code is unreachable.
-  clientRenderBoundary();
-  completeBoundaryWithStyles();
-  completeBoundary();
-  completeSegment();
-});
+// This runtime may be sent to the client multiple times (if FizzServer.render
+//  is called more than once). Here, we check whether the mutation observer
+//  was already created / installed
+if (!window.$REACT_FIZZ_OBSERVER) {
+  // TODO: Eventually remove, we currently need to set these globals for
+  // compatibility with ReactDOMFizzInstructionSet
+  window.$RC = completeBoundary;
+  window.$RM = new Map();
+  window.$REACT_FIZZ_OBSERVER = new MutationObserver(mutations => {
+    mutations.forEach(mutation => {
+      mutation.addedNodes.forEach(handleNode);
+    });
+  });
+  // $FlowFixMe[incompatible-call] document.body should exist at this point
+  window.$REACT_FIZZ_OBSERVER.observe(document.body, {
+    childList: true,
+    subtree: true,
+  });
+
+  const existingNodes = document.getElementsByTagName('div');
+  for (let i = 0; i < existingNodes.length; i++) {
+    handleNode(existingNodes[i]);
+  }
+}
+
+function handleNode(node_ /*: Node */) {
+  if (node_.nodeType !== 1) {
```

Comment:
Sorry, I'm not quite following here  (just want to understand) 
How would we know what to expect for parentNode / would this be the embedding context? (For example, some of our tests render into div containers).

Reply:
I was just thinking `document.body`. Our support for embedding a stream inside a tree is a little poorly supported anyway. It doesn't really make much sense to delay the end tags. It was kind of just done in case someone wanted to try it.

I think going forward our suggestion for rendering React only to part of the tree will be an API that treats and existing ID as the target for the shell but assume that the streamed content gets concatenated to the body.

We can start by making this option only work for "body". Otherwise you have to use the inline scripts.

I think that the way we should do this though is to narrow the scope of the `MutationObserver` option. My concern is that right now it affects everything in the whole document which will call this a lot on a large document and risks more collisions.

In other words, let's just remove the `subtree: true` option and only observe injected nodes in the root.